### PR TITLE
PR gateway: Added dependency between the groups to run in parallel/sequence

### PR DESCRIPTION
**Pull request by kparichay**
Working on issue #485

This commit introduces dependency between the various groups.
If the dependency is set, failure of one group will cause other groups to skip
If the dependency is not set, all the groups will run in parallel

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: ci/taos/checker-pr-audit-async.sh [hunk 1]
```diff
@@ -322,16 +322,19 @@ if [[ $global_check_result == "success" ]]; then
 
     # If contributors want later, let's inform developers of CI test result to go to a review process as a final step before merging a PR
     echo "[DEBUG] All audit modules are passed - it is ready to review! :shipit:. Note that CI bot has two sub-bots such as TAOS/pr-audit-all and TAOS/pr-format-all."
+    exit 0
```
Comment (leemgs) [inline ci/taos/checker-pr-audit-async.sh]: We need to move the existing  Line 337-340 in front of the if statement if you want to append exit xxx.

File: ci/taos/checker-pr-gateway.sh [hunk 2]
```diff
@@ -135,34 +135,70 @@ $checkout_branch
...
+    pushd ./taos/
+    ${checker_list[$i]} $checker_args                                    | tee -a $logfile &
+    local pid=$!
+    popd
+    echo -e "[DEBUG] Running..."                                         | tee -a $logfile
+
+    # if the dependency between groups is to be enforced and current group fails, skip remaining groups
+    if [[ $dependency_policy_between_groups == 1 ]]; then
+      wait $pid
+      if [[ $? != 0 ]]; then
```
Comment (leemgs) [inline ci/taos/checker-pr-gateway.sh]: Looks good.

File: ci/taos/checker-pr-gateway.sh [hunk 3]
```diff
@@ -135,34 +135,70 @@ $checkout_branch
...
+
+# Ready a checker: audit
+checker_list+=("./checker-pr-audit-async.sh")
+checker_name_list+=("audit")
+checker_logfile_list+=("${dir_ci}/${dir_commit}/checker-pr-audit.log")
+echo -e "[DEBUG] Added ${checker_name_list[-1]} checker"
+
+# Run all the checkers
+echo -e "[DEBUG] Running all checkers"
+run_all_checkers checker_list checker_name_list checker_logfile_list $1 $2 $3 $4 $5 $6
```
Comment (leemgs) [inline ci/taos/checker-pr-gateway.sh]: For readability and maintenance, 
`s/checker_list/checker_cmd_list/g `

File: ci/taos/checker-pr-gateway.sh [hunk 4]
```diff
@@ -135,34 +135,70 @@ $checkout_branch
...
+  done
+}
+
+# --------------------------- Create and run checkers -------------------------------------
+# Create an empty list for the checkers to be run with the corresponding elements
+# The order of the checkers in the list creates a dependency between them determining the order in which they will run
+# 1) Checker itself to be run
+# 2) Name of the checker
+# 3) Log file to log the output of the checker
+checker_list=()
```
Comment (leemgs) [inline ci/taos/checker-pr-gateway.sh]: Witch better? Please consider a maintenance in order that contributors can easily understand and modify.
1. checker_list=()
2. declare -a checker_list

File: ci/taos/checker-pr-gateway.sh [hunk 5]
```diff
@@ -135,34 +135,70 @@ $checkout_branch
...
+checker_name_list=()
+checker_logfile_list=()
+
+# Ready a checker: format
+checker_list+=("./checker-pr-format-async.sh")
+checker_name_list+=("format")
+checker_logfile_list+=("${dir_ci}/${dir_commit}/checker-pr-format.log")
+echo -e "[DEBUG] Added ${checker_name_list[-1]} checker"
```
Comment (leemgs) [inline ci/taos/checker-pr-gateway.sh]: Is it correct? do we have to use -1 without 0?
`${checker_name_list[-1]} `

Comment (kparichay) [reply ci/taos/checker-pr-gateway.sh]: "If the subscript used to reference an element of an indexed array evaluates to a number less than zero, it is interpreted as relative to one greater than the maximum index of the array, so negative indices count back from the end of the array, and an index of -1 refers to the last element."

- From https://www.gnu.org/software/bash/manual/bashref.html#Arrays

File: ci/taos/checker-pr-gateway.sh [hunk 6]
```diff
@@ -135,34 +135,70 @@ $checkout_branch
...
+#  arg2: list of checker's name
+#  arg3: list of checker's log file
+#  arg@: all remaining arguments are to be passed to the checkers (common for all the checkers)
+function run_all_checkers(){
+  local -n checker_list=$1
+  local -n name_list=$2
+  local -n logfile_list=$3
+  local checker_args=${@:4}
+
+  local num_checkers=${#checker_list[@]}
```
Comment (leemgs) [inline ci/taos/checker-pr-gateway.sh]: 👍

File: ci/taos/checker-pr-audit-async.sh (hunk 1 above)
Comment (kparichay) [reply ci/taos/checker-pr-audit-async.sh]: Updated.

File: ci/taos/checker-pr-gateway.sh (hunk 3 above)
Comment (kparichay) [reply ci/taos/checker-pr-gateway.sh]: Done.

File: ci/taos/checker-pr-gateway.sh (hunk 4 above)
Comment (kparichay) [reply ci/taos/checker-pr-gateway.sh]: Updated to `declare -a checker_list`

**Comment by kparichay** [comment]
**V2**
updated `checkers_list` to `checkers_cmd_list`
using `declare -a` to declare arrays
using `exit_code` variable and exiting at the end of a checker

**Comment by leemgs** [comment]
@kparichay Please update your PR one more as a final step. :)

**Comment by kparichay** [comment]
Updated.

**Comment by leemgs** [comment]
* https://github.com/nnsuite/TAOS-CI/issues/492

**Pull request merged**